fastcrypto/src/secp256r1.rs

```diff
@@ -0,0 +1,492 @@
+//! ```
+
+use crate::{
+    encoding::{Base64, Encoding},
+    error::FastCryptoError,
+    pubkey_bytes::PublicKeyBytes,
+    serde_helpers::keypair_decode_base64,
+    traits::{
+        AllowedRng, Authenticator, EncodeDecodeBase64, KeyPair, SigningKey, ToFromBytes,
+        VerifyingKey,
+    },
+};
+use fastcrypto_derive::{SilentDebug, SilentDisplay};
+use once_cell::sync::OnceCell;
+use p256::ecdsa::Signature as ExternalSignature;
+use p256::ecdsa::SigningKey as ExternalSecretKey;
+use p256::ecdsa::VerifyingKey as ExternalPublicKey;
+use p256::elliptic_curve::group::GroupEncoding;
+
+use crate::hash::HashFunction;
+use crate::hash::Sha256;
+use generic_array::GenericArray;
+use p256::elliptic_curve::bigint::ArrayEncoding;
+use p256::elliptic_curve::ops::Reduce;
+use p256::elliptic_curve::{Curve, DecompactPoint};
+use p256::{AffinePoint, NistP256, ProjectivePoint, Scalar, U256};
+use serde::{de, Deserialize, Serialize};
+use signature::{Signature, Signer, Verifier};
+use std::{
+    fmt::{self, Debug, Display},
+    str::FromStr,
+};
+use zeroize::Zeroize;
+
+pub const PUBLIC_KEY_SIZE: usize = 33;
+pub const PRIVATE_KEY_SIZE: usize = 32;
+pub const SIGNATURE_SIZE: usize = 64;
+
+/// Secp256r1 public key.
+#[readonly::make]
+#[derive(Debug, Clone)]
+pub struct Secp256r1PublicKey {
+    pub pubkey: ExternalPublicKey,
+    pub bytes: OnceCell<[u8; PUBLIC_KEY_SIZE]>,
+}
+
+/// Binary representation of an instance of [Secp256r1PublicKey].
+pub type Secp256r1PublicKeyBytes = PublicKeyBytes<Secp256r1PublicKey, { PUBLIC_KEY_SIZE }>;
+
+/// Secp256r1 private key.
+#[readonly::make]
+#[derive(SilentDebug, SilentDisplay)]
+pub struct Secp256r1PrivateKey {
+    pub privkey: ExternalSecretKey,
+    pub bytes: OnceCell<[u8; PRIVATE_KEY_SIZE]>,
+}
+
+/// Secp256r1 signature.
+#[readonly::make]
+#[derive(Debug, Clone)]
+pub struct Secp256r1Signature {
+    pub sig: ExternalSignature,
+    pub bytes: OnceCell<[u8; SIGNATURE_SIZE]>,
+}
+
+impl std::hash::Hash for Secp256r1PublicKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.as_ref().hash(state);
+    }
+}
+
+impl PartialOrd for Secp256r1PublicKey {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.as_ref().partial_cmp(other.as_ref())
+    }
+}
+
+impl Ord for Secp256r1PublicKey {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.as_ref().cmp(other.as_ref())
+    }
+}
+
+impl PartialEq for Secp256r1PublicKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.pubkey == other.pubkey
+    }
+}
+
+impl Eq for Secp256r1PublicKey {}
+
+impl VerifyingKey for Secp256r1PublicKey {
+    type PrivKey = Secp256r1PrivateKey;
+    type Sig = Secp256r1Signature;
+    const LENGTH: usize = PUBLIC_KEY_SIZE;
+}
+
+impl Verifier<Secp256r1Signature> for Secp256r1PublicKey {
+    fn verify(&self, msg: &[u8], signature: &Secp256r1Signature) -> Result<(), signature::Error> {
```

Comment:
for secp256k1, we have the recovery id appended to the 64 bytes signature. perhaps we can do the same for the r1 signature as well?